restapi/datasource_api_object.go

```diff
@@ -115,34 +131,75 @@ func dataSourceRestApiRead(d *schema.ResourceData, meta interface{}) error {
     } /* End Loop through parts */
 
     data_array = (*ptr).([]interface{})
-  } else {
+  } else if "" == results_key && !single_obj {
     data_array = result.([]interface{})
+  } else { // single_obj
+    ptr := &result
+    single_object_data = (*ptr).(map[string]interface{})
   }
 
-  /* Loop through all of the results seeking the specific record */
-  for _, item := range data_array {
-    hash := item.(map[string]interface{})
-
-    /* We found our record */
-    if hash[search_key] == search_val {
-      id = hash[id_attribute].(string)
+  if !single_obj {
+
+    /* Loop through all of the results seeking the specific record */
+    for _, item := range data_array {
+      hash := item.(map[string]interface{})
+
+      // Parse search_key
+      search_parts := strings.Split(search_key, "/")
+      search_part := ""
+      search_seen := ""
+      search_hash := hash
+
+      // Loop through search parts
+      for len(search_parts) > 1 {
+        log.Printf("datasource_api_object.go: Looping through search_parts\n")
+        search_part, search_parts = search_parts[0], search_parts[1:]
+
+        // Protect against double slashes by mistake
+        if "" == search_part { break }
+
+        if _, ok := search_hash[search_part]; ok {
+          log.Printf("search_hash[search_part] exists: '%s'\n", search_hash[search_part])
+          search_hash = search_hash[search_part].(map[string]interface{})
+          search_seen += "/" + search_part
+        } else {
+          log.Printf("Failed to find %s in returned data structure after finding '%s'", search_part, search_seen)
+        }
+      } // end search_parts loop
+
+      search_part, search_parts = search_parts[0], search_parts[1:]
+      search_data_map := search_hash
+      log.Printf("search_part is set to '%s'\n", search_part)
+      if search_data_map[search_part] == search_val {
+        /* We found our record */
+        log.Printf("datasource_api_object.go: Found our record\n")
+        id = search_data_map[id_attribute].(string)
+        /* But there is no id attribute??? */
+        if "" == id {
+          log.Printf("datasource_api_object.go: But record did not have '%s' attribute\n", id_attribute)
+          return(errors.New(fmt.Sprintf("The object for '%s'='%s' did not have the id attribute '%s'", search_key, search_val, id_attribute)))
+        }
+        break
+      } // end if for checking search_key against search_val
+    } // end data_array loop
```

Comment:
There should be a break here if the object was already found. If we found our item after the second of 500 loops, we should avoid the 498 subsequent loops.

I suggest adding a `found` boolean after line 145, setting it to true in the line after 182 and then checking for truth here after line 183.

Reply:
I don't think an extra break is needed.  The break that is there already does stop the data_array loop if a matching object is found. The break on line 182 in the comment (which is now line 173 in newer commits) is the same as what you have on line 134 of your version. Note that the only loop that this break is inside of is the data_array loop.  Perhaps you were thinking that it was inside my new `for len(search_parts) > 1 {}` loop?

Reply:
Yep - you are correct. 173 is a break against the loop for all results and I confused myself as I read the diff (thinking it was inside ` for len(search_parts) > 1`. Looking at the file instead of the diff made it more clear. Sorry for the noise.